pkg/webhook/webhook.go

```diff
@@ -0,0 +1,345 @@
+	if obj.Annotations[IsDefaultSnapshotClassAnnotation] == "true" {
+		return true
+	}
+
+	return false
+}
+
+func getDefaultSnapshotClassName(driver string, snapClient clientset.Interface) (string, error) {
+	list, err := snapClient.VolumesnapshotV1alpha1().VolumeSnapshotClasses().List(metav1.ListOptions{})
+	if err != nil {
+		return "", err
+	}
+	defaultClasses := []snapshotv1alpha1.VolumeSnapshotClass{}
+
+	for _, class := range list.Items {
+		if IsDefaultAnnotation(class.ObjectMeta) && class.Snapshotter == driver {
+			defaultClasses = append(defaultClasses, class)
+			glog.V(5).Infof("get defaultClass added: %s", class.Name)
+		}
+	}
+	if len(defaultClasses) == 0 {
+		return "", fmt.Errorf("cannot find default snapshot class")
+	}
+	if len(defaultClasses) > 1 {
+		glog.V(4).Infof("get DefaultClass %d defaults found", len(defaultClasses))
+		return "", fmt.Errorf("%d default snapshot classes were found", len(defaultClasses))
+	}
+	return defaultClasses[0].Name, nil
+}
+
+func getSnapshotClassNameFromContent(content *snapshotv1alpha1.VolumeSnapshotContent, snapClient clientset.Interface) (string, error) {
+	if content.Spec.VolumeSnapshotClassName != nil {
+		return *content.Spec.VolumeSnapshotClassName, nil
+	}
+	if content.Spec.VolumeSnapshotSource.CSI == nil {
+		return "", fmt.Errorf("VolumeSNapshotContent does not has CSI volume source")
+	}
+	return getDefaultSnapshotClassName(content.Spec.VolumeSnapshotSource.CSI.Driver, snapClient)
+}
+
+// validateSnapshot checks whether the values set on the VolumeSnapshot object are valid.
+// also set the snapshot class if it is not specified
+func validateSnapshot(snapshot *snapshotv1alpha1.VolumeSnapshot, clientset kubeclientset.Interface, snapClient clientset.Interface) (*snapshotv1alpha1.VolumeSnapshot, bool, error) {
+	if snapshot.Spec.Source != nil {
+		source := snapshot.Spec.Source
+		if len(source.Name) == 0 {
+			return nil, false, fmt.Errorf("Snapshot.Spec.Source.Name can not be empty")
+		} else if !supportedSourceKinds.Has(string(source.Kind)) {
+			return nil, false, fmt.Errorf("Snapshot.Spec.Source.Kind exepct %v, got %q", supportedSourceKinds, source.Kind)
+		}
+		pvc, err := validateVolume(snapshot, clientset)
+		if err != nil {
+			return nil, false, err
+		}
+		exist, driver, err := validateSnapshotClass(snapshot, pvc, clientset, snapClient)
+		if err != nil {
+			return nil, false, err
+		}
+		if exist {
+			return snapshot, false, nil
+		}
+		defaultClassName, err := getDefaultSnapshotClassName(driver, snapClient)
+		if err != nil {
+			return nil, false, err
+		}
+		snapshotClone := snapshot.DeepCopy()
+		snapshotClone.Spec.VolumeSnapshotClassName = &defaultClassName
+		return snapshotClone, true, nil
+	}
+
+	// if source is not specified, the content name must be set
+	if snapshot.Spec.SnapshotContentName == "" {
+		return nil, false, fmt.Errorf("Cannot set Snapshot.Spec.Source to nil and snapshot.Spec.SnapshotContentName to empty at the same time.")
```

Comment:
lower case to start an error: s/Cannot/cannot

Reply:
fixed